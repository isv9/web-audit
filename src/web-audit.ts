import { auditBlockSemantics, auditCommonSemantics, auditHeaderSemantics, auditTextSemantics } from './audits/semantic'
import { auditLinks } from './audits/links'

export type WebDocument = {
  getElementsByTagNameCount (tag: string): number;
  querySelectorAll (query: string): { length: number };
  getEmptyElementsByTagName (tag: string): { length: number }
};

type AuditSectionResult = {
  name: string;
  auditResults: AuditResult[];
};

export type AuditResult = {
  name?: string;
  tables: (AuditResultTable)[];
  liveCollections?: (AuditResultLiveCollection)[];
  logs?: string[];
  warnings?: string[];
  errors?: string[];
};

type AuditResultTable = { name?: string } & { [key: string]: string | number };
export type AuditResultLiveCollection = { name: string, collection: object };

export class WebAudit {

  private readonly webDocument: WebDocument

  constructor (webDocument: WebDocument) {
    this.webDocument = webDocument
  }

  audit () {
    const auditSemanticsResult = this.auditSemantics()
    const auditLinksResult = this.auditLinks()
    console.group('web audit')
    WebAudit.renderAuditSectionResult(auditSemanticsResult)
    WebAudit.renderAuditSectionResult(auditLinksResult)
    console.groupEnd()
  }

  private auditLinks (): AuditSectionResult {
    return {
      name: 'links',
      auditResults: [
        auditLinks(this.webDocument)]
    }
  }

  private auditSemantics (): AuditSectionResult {

    return {
      name: 'semantics',
      auditResults: [auditCommonSemantics(this.webDocument),
        auditHeaderSemantics(this.webDocument),
        auditTextSemantics(this.webDocument),
        auditBlockSemantics(this.webDocument)]
    }
  }

  static renderAuditSectionResult (auditSectionResult: AuditSectionResult) {
    const { name, auditResults } = auditSectionResult
    const details = {
      errorsCount: 0,
      warningsCount: 0
    }
    auditResults.forEach(auditResult => {
      const errorsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'errors')
      const warningsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'warnings')
      details.errorsCount += errorsCount
      details.warningsCount += warningsCount
    })
    console.group(`${name}, (errors="${details.errorsCount}", warnings="${details.warningsCount}")`)
    auditResults.forEach(auditResult => WebAudit.renderAuditResult(auditResult))
    console.groupEnd()
  }

  static renderAuditResult (auditResult: AuditResult) {
    const { name, tables, errors = [], warnings = [], liveCollections = [], logs = [] } = auditResult
    const errorsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'errors')
    const warningsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'warnings')
    console.group(`${name}, (errors="${errorsCount}", warnings="${warningsCount}")`)
    tables.forEach(({ name, nodes, ...tableDataSet }) => {
      console.log(name)
      console.table(tableDataSet)
    })
    liveCollections.forEach(({ name, collection }) => {
      console.log(name)
      console.log(collection)
    })
    logs.forEach(message => console.log(message))
    console.log('Summary')
    if (errorsCount > 0 || warningsCount > 0) {
      errors.forEach(message => console.error(message))
      warnings.forEach(message => console.warn(message))
    } else {
      console.log('It is OK')
    }

    console.groupEnd()
  }

  static getAuditResultMessagesCount (auditResult: AuditResult, key: keyof AuditResult): number {
    let property = auditResult[key]
    if (property === null || property === undefined) {
      property = []
    }
    if (Array.isArray(property)) {
      return property.length
    }
    throw new Error('Property type must be array')
  }
}

