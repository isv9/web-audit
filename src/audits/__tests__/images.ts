import { WebDocument } from '../../web-audit'
import { auditImages } from '../images'

describe('auditImages', () => {
  it('document is ok', () => {
    const document: Pick<WebDocument, 'querySelectorAll' | 'getElementsByTagNameCount'> = {
      getElementsByTagNameCount: jest.fn((tag) => tag === 'image' ? 0 : 3),
      querySelectorAll: jest.fn(() => ({ length: 1 })),
    }
    expect(auditImages(document)).toMatchSnapshot()
  })

  it('document has image tag', () => {
    const document: Pick<WebDocument, 'querySelectorAll' | 'getElementsByTagNameCount'> = {
      getElementsByTagNameCount: jest.fn((tag) => tag === 'image' ? 1 : 3),
      querySelectorAll: jest.fn(() => ({ length: 1 })),
    }
    expect(auditImages(document)).toMatchSnapshot()
  })

  it('document has not flexible images', () => {
    const document: Pick<WebDocument, 'querySelectorAll' | 'getElementsByTagNameCount'> = {
      getElementsByTagNameCount: jest.fn((tag) => tag === 'img' ? 10 : 0),
      querySelectorAll: jest.fn(() => ({ length: 0 })),
    }
    expect(auditImages(document)).toMatchSnapshot()
  })

})
