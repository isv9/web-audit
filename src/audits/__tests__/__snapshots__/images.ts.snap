// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auditImages document has image tag 1`] = `
Object {
  "errors": Array [
    "Tag image is deprecated",
  ],
  "logs": Array [
    "Document has 1 img with srcset attribute",
    "Document has some flexible images",
  ],
  "name": "images in dom",
  "tables": Array [
    Object {
      "image": 1,
      "img": 3,
      "map": 3,
      "picture": 3,
      "source": 3,
      "video": 3,
    },
  ],
}
`;

exports[`auditImages document has not flexible images 1`] = `
Object {
  "errors": Array [
    "Document does not have any flexible images",
  ],
  "logs": Array [
    "Document has 0 img with srcset attribute",
  ],
  "name": "images in dom",
  "tables": Array [
    Object {
      "image": 0,
      "img": 10,
      "map": 0,
      "picture": 0,
      "source": 0,
      "video": 0,
    },
  ],
}
`;

exports[`auditImages document is ok 1`] = `
Object {
  "errors": Array [],
  "logs": Array [
    "Document has 1 img with srcset attribute",
    "Document has some flexible images",
  ],
  "name": "images in dom",
  "tables": Array [
    Object {
      "image": 0,
      "img": 3,
      "map": 3,
      "picture": 3,
      "source": 3,
      "video": 3,
    },
  ],
}
`;
